style: add .editorconfig and reformat js file

ert78gb opened the pull request:
use 2 space indent in every file

csantanapr commented:
Did you already submitted an ICLA to Apache? https://github.com/apache/incubator-openwhisk-client-js/blob/master/CONTRIBUTING.md#contributor-license-agreement

ert78gb commented:
@csantanapr 
- added `npm run lint` to travis.yml
- added pre-commit hook to the project and add a few word to the contribution doc
- reformat code to match eslint rules

When I added eslintrc.json I checked your coding style, but I think would be better if we use a more common style as [standard](https://github.com/standard/standard)

I have ICLA as Róbert Kiss

csantanapr commented:
@ert78gb I see some conflicts can you take a look?

>When I added eslintrc.json I checked your coding style, but I think would be better if we use a more common style as standard

I'm OK to use [standard](https://github.com/standard/standard)

ert78gb commented:
I prefer the standard style. I tried to use those settings which modifies the current code style the least. I will check the conflict

csantanapr commented:
I heard good things about standard so it should be OK

ert78gb commented:
@csantanapr I finished applying the new style rules

csantanapr commented on README.md:
```diff
@@ -66,19 +67,19 @@ ow.actions.invoke('sample').then(result => console.log(result))
 
 - **apihost.** Hostname and optional port for openwhisk platform, e.g. `openwhisk.ng.bluemix.net` or `my_whisk_host:80`. Used with API URL template `${protocol}://${apihost}/api/v1/`. If port is missing or port value is 443 in the apihost string, protocol is HTTPS. Otherwise, protocol is HTTP.
-- **api_key.** Authorisation key for user account registered with OpenWhisk platform.
+- **apiKey.** Authorisation key for user account registered with OpenWhisk platform.
```
Please do not remove the constructor API parameters this will be braking API change. 
Find a way in the linter or code style rule to exclude this 

we need to keep the parameter names. It’s ok to add new aliases for the new names and mark the current ones deprecated

ert78gb replied:
ok I will do

ert78gb replied:
I added "old style" options parameters and marked them as deprecated. I did not change the readme file, because the new users use the new parameters :)

jthomas replied:
I don't think we should be marking the current parameters as `deprecated` to fit arbitrary style rules. Changing the internal code style is fine but doing this to the constructor parameters is unnecessary and confusing. I believe that underscores are more common in JavaScript-land and much less error-prone. It's very easy to miss a capitalisation when reading documentation.

csantanapr commented:
Thanks for the effort

jthomas commented on README.md:
```diff
@@ -66,19 +67,19 @@ ow.actions.invoke('sample').then(result => console.log(result))
+- **ignoreCerts**. Turns off server SSL/TLS certificate verification. This allows the client to be used against local deployments of OpenWhisk with a self-signed certificate. Defaults to false.
+- **apigwToken**. API Gateway service authentication token. This is mandatory for using an external API Gateway service, rather than the built-in api gateway.
+- **apigwSpaceGuid**. API Gateway space identifier. This is optional when using an API gateway service, defaults to the authentication uuid.
 
```
I don't think we should be changing these parameters. Moving to another arbitrary style for the constructor parameters does not add any value. I think that "underscores" are more common in JS and it's less error-prone than capitalisation. It's easy to miss a capital letter when you are reading documentation.

ert78gb commented:
I think the camelCase more commom naming convention, but it’s up to u. You are the owner of the project you have to make a decision, but it should have be the first.

ert78gb commented:
What is the next step with this PR? It is a big PR changed many files. I will not resolve the conflicts sry for it, but I did twice. Pls make a decision do you accept it or close. Every OSS software has a style guide with lintert, editorconfig, etc.

csantanapr commented:
There is a lot of value on this PR
@jthomas what about as a compromised we keep both ways as supported properties for the contractor that user is free underscore or camel case when in the constructor when initialize 
Would this option work for both of you as a compromised?
Then the only thing left in this PR is to update readme to day that properties names are both supported for the end user and resolve the conflict.

csantanapr commented:
If I would have my way I would name the properties single words “host” “key” “insecure” “token” “space”

ert78gb commented:
@csantanapr I think this is the worst choice use both syntax. It made over complicate the code, need to write more unit test if adding new options parameter. The second way is looks better :)

csantanapr commented:
Then @ert78gb let’s keep the properties with underscore they way it is today

ert78gb commented:
@csantanapr just for the clarification. Would you like to use the underscore naming convention int the client constructor options parameters or everywhere in the code?

csantanapr commented:
Only in the client contructor options

jthomas commented:
@ert78gb This PR adds a lot of value to the project and I definitely want to merge it. Thanks for the on-going work on this. 

I'm happy with all the changes, apart from changing the client constructor parameters to use camelcase. Changing the constructor parameters would break existing code and require us to update all documentation or examples using these parameters. Supporting both styles is unnecessary for the reasons you pointed out. 

If you can make a final change to revert the client constructor parameter changes, this LGTM for merging. 💯

csantanapr commented on README.md:
````diff
@@ -57,7 +59,7 @@ exports.main = action
 ```javascript
 var openwhisk = require('openwhisk');
-var options = {apihost: 'openwhisk.ng.bluemix.net', apiKey: '...'};
+var options = {apihost: 'openwhisk.ng.bluemix.net', api_key: '...'};
````
Thanks for the update

csantanapr commented on README.md:
```diff
@@ -67,19 +69,19 @@ ow.actions.invoke('sample').then(result => console.log(result))
 
 - **apihost.** Hostname and optional port for openwhisk platform, e.g. `openwhisk.ng.bluemix.net` or `my_whisk_host:80`. Used with API URL template `${protocol}://${apihost}/api/v1/`. If port is missing or port value is 443 in the apihost string, protocol is HTTPS. Otherwise, protocol is HTTP.
-- **apiKey.** Authorisation key for user account registered with OpenWhisk platform.
+- **api_key.** Authorisation key for user account registered with OpenWhisk platform.
```
Thanks for the update

csantanapr commented on README.md:
```diff
@@ -67,19 +69,19 @@ ow.actions.invoke('sample').then(result => console.log(result))
-- **apigwSpaceGuid**. API Gateway space identifier. This is optional when using an API gateway service, defaults to the authentication uuid.
+- **ignore_certs**. Turns off server SSL/TLS certificate verification. This allows the client to be used against local deployments of OpenWhisk with a self-signed certificate. Defaults to false.
+- **apigw_token**. API Gateway service authentication token. This is mandatory for using an external API Gateway service, rather than the built-in api gateway.
+- **apigw_space_guid**. API Gateway space identifier. This is optional when using an API gateway service, defaults to the authentication uuid.
```
Thanks for the update

ert78gb commented:
Hi @csantanapr and @jthomas ,

First of all, sorry for my late response but I was extremely busy in the last few month. 

I reverted the constructor parameters to the old style and the documentation too.

I suggest to increase major version number when you release a new version, because if someone use "private" method of an object it may cause breaks. Example if someone used the lib/action.js -> action_body method it renamed to actionBody. 

I have a few other refactoring suggestion. Example: use pure function instead if object method and the method use only input parameters of the method lib/action.js -> actionBody with this technic you could minimalized the api exposition

Happy holidays,
Robi

jthomas commented:
👋 @ert78gb. Thanks for the updates to this PR, everything looks perfect to me now 💯. If you have other re-factoring suggestions, feel free to open PRs. FYI internal re-factoring is great but anything that changes the external API for cosmetic reasons is not likely to be considered for merging if that makes sense?

@csantanapr Is your review now resolved?

I agree with @csantanapr that this won't require a major version change as semver refers to the client libraries public API. I know in JS land people often use private methods but that's not something we manage through the semver version.

csantanapr commented:
@jthomas go ahead and merge if you are OK
